lib/puppet/defaults.rb

```diff
@@ -71,12 +71,14 @@ def self.default_vendormoduledir
         :type    => :boolean,
         :desc    => 'Whether to enable a pre-Facter 4.0 release of Facter (distributed as
           the "facter-ng" gem). This is not necessary if Facter 3.x or later is installed.
-          This setting is still experimental and has been only included on Windows builds',
+          This setting is still experimental.',
         :hook    => proc do |value|
-                      if value && Puppet::Util::Platform.windows?
+                      if value
                         begin
+                          Gem::Specification::find_by_name('facter-ng', '>= 4.0.0', '< 5')
+                          Object.send(:remove_const, :Facter) if defined?(Facter)
                           require 'facter-ng'
-                        rescue LoadError
+                        rescue Gem::MissingSpecError
                           raise ArgumentError, 'facter-ng could not be loaded'
```

Comment:
if `facter-ng` is not available, then we're going to raise ArgumentError. So why check if there is a gem specification first?

Reply:
The check for `facter-ng` needs to be done before trying to load `facter-ng` in order to not remove `Facter` if `facter-ng` is not installed(because Facter and FacterNg implement the same module). If `Gem::Specification` does not raise error than `require 'facter-ng'` should work fine. 
The raise for `ArgumentError` in case `facter-ng` is not found is just to have a more concise error message.

Reply:
Is it possible for `find_by_name` to return success, but `require 'facter-ng'` to fail? I'm thinking of perhaps a dependency of `facter-ng` can't be loaded, like missing `ffi` or something? In which case it might be better to save off the `Facter` constant and add it back if `require 'facter-ng'` fails?

```ruby
begin
  ORIGINAL_FACTER = Object.const_get(:Facter)
  Object.send(:remove_const, :Facter)
  require 'facter-ng'
rescue LoadError
  const_set(:Facter, ORIGINAL_FACTER)
  raise ArgumentError, 'facter-ng could not be loaded'
end
```

Reply:
It's unlikely in the puppet-agent context, and the facter-ng that we use as a gem for puppet gem does not have dependencies on gems with native extension. But the approach with storing the original facter handles all the cases so i'm for that.